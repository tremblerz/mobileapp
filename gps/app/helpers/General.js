import AsyncStorage from '@react-native-community/async-storage';
import DocumentPicker from 'react-native-document-picker';

/**
 * Get data from store
 *
 * @param {string} key
 * @param {boolean} isString
 */
export async function GetStoreData(key, isString = true) {
  try {
    let data = await AsyncStorage.getItem(key);

    if (isString) {
      return data;
    }

    return JSON.parse(data);
  } catch (error) {
    console.log(error.message);
  }
  return false;
}

/**
 * Set data from store
 *
 * @param {string} key
 * @param {object} item
 */
export async function SetStoreData(key, item) {
  try {
    //we want to wait for the Promise returned by AsyncStorage.setItem()
    //to be resolved to the actual value before returning the value
    if (typeof item !== 'string') {
      item = JSON.stringify(item);
    }

    return await AsyncStorage.setItem(key, item);
  } catch (error) {
    console.log(error.message);
  }
}

/**
 * Remove data with associated key from store
 *
 * @param {string} key
 * @param {object} item
 */
export async function RemoveStoreData(key) {
  try {
    return await AsyncStorage.removeItem(key);
  } catch (error) {
    console.log(error.message);
  }
}

export async function pickFile() {
  // Pick a single file - returns actual path on Android, file:// uri on iOS
  try {
    const res = await DocumentPicker.pick({
      type: [DocumentPicker.types.zip, DocumentPicker.types.allFiles],
      usePath: true,
    });
    return res.uri;
  } catch (err) {
    if (DocumentPicker.isCancel(err)) {
      // User cancelled the picker, exit any dialogs or menus and move on
    } else {
      throw err;
    }
  }
}
