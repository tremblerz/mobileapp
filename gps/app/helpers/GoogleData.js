/**
 * Rounds float number to a desired number of decimal places and returns a float
 * number. NOTE: .toFixed() returns a string, but number is required.
 * @param num - number
 * @param digits - amount of digits to round
 * @returns {number}
 */
function toFixedNumber(num, digits) {
  const pow = Math.pow(10, digits);
  return Math.round(num * pow) / pow;
}

/**
 * Formats a provided google placeVisit to a local format making sure
 * float numbers have constant number of decimal places as float numbers
 * has to be exact for later comparison.
 *
 * @param placeVisit - google place object
 * @returns {{latitude: number, time: string, longitude: number}}
 */
function formatLocation(placeVisit) {
  return {
    latitude: toFixedNumber(placeVisit.location.latitudeE7 * 10 ** -7, 7),
    longitude: toFixedNumber(placeVisit.location.longitudeE7 * 10 ** -7, 7),
    time: placeVisit.duration.startTimestampMs,
  };
}

export function extractLocations(googleLocationHistory) {
  return (googleLocationHistory?.timelineObjects || []).map(location => {
    // Only import visited places, not paths for now
    if (location?.placeVisit) {
      return formatLocation(location.placeVisit);
    }
  });
}
