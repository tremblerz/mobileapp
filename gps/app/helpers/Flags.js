import { assign } from 'lodash';
import React, { useContext, useEffect, useState } from 'react';

import { getBuildtimeFlags } from '../constants/flagsEnv';
import { FEATURE_FLAG_VALS } from '../constants/storage';
import { GetStoreData, SetStoreData } from './General';

/**
 * Converts the flag name from env into a more human readable form.
 *
 * Example:
 *
 * `getCleanedFlagName('hello_world')` becomes `Hello world`
 *
 */
export const getCleanedFlagName = name => {
  const withSpaces = name.replace('_', ' ');
  return (
    withSpaces.charAt(0).toUpperCase() + withSpaces.slice(1, withSpaces.length)
  );
};

export const FlagsContext = React.createContext([
  getBuildtimeFlags(),
  () => {},
]);

/**
 * Overwrites properties of `oldFlags` with properties from `newFlags`,
 * omitting any properties from `newFlags` that don't exist on `oldFlags`
 *
 * Examples:
 *
 * `mergeFlags({feature1: false}, {feature1: true})` becomes `{feature1: true}`
 *
 * `mergeFlags({feature1: false}, {feature1: true, feature2: true})` becomes `{feature1: true}`
 */
export const mergeFlags = (oldFlags, newFlags) => {
  return assign(oldFlags, newFlags, (objectValue, sourceValue) => {
    if (objectValue) {
      return sourceValue;
    }
  });
};

export const FlagsProvider = ({ children }) => {
  const buildTimeFlags = getBuildtimeFlags();
  const [flags, setFlags] = useState(buildTimeFlags);

  useEffect(() => {
    const getInitalFlagVals = async () => {
      const storedFlags = await GetStoreData(FEATURE_FLAG_VALS, false);

      if (storedFlags) {
        // Overwrite existing properties of `getBuildtimeFlags` with stored values from async storage,
        // omitting any stored value that is not present on `getBuildtimeFlags`.
        const initialFlags = mergeFlags(buildTimeFlags, storedFlags);
        setFlags(initialFlags);
        await SetStoreData(FEATURE_FLAG_VALS, initialFlags);
      }
    };

    getInitalFlagVals();
  }, [buildTimeFlags]);

  return (
    <FlagsContext.Provider value={[flags, setFlags]}>
      {children}
    </FlagsContext.Provider>
  );
};

/**
 * Custom hook that can be used to fetch the list of all flags,
 * and to set the value for a particular flag.
 */
export function useFlags() {
  const [flags, setFlags] = useContext(FlagsContext);

  const setFlag = async (flagName, flagVal) => {
    const newFlags = { ...flags, [flagName]: flagVal };
    setFlags(newFlags);
    await SetStoreData(FEATURE_FLAG_VALS, newFlags);
  };

  return [flags, setFlag];
}
