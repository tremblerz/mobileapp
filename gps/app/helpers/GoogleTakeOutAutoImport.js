import dayjs from 'dayjs';
import { NativeModules, Platform } from 'react-native';
import RNFS from 'react-native-fs';
/**
 * Checks the download folder, unzips and imports all data from Google TakeOut
 */
import { subscribe, unzip } from 'react-native-zip-archive';

import { extractLocations } from '../helpers/GoogleData';

export class NoRecentLocationsError extends Error {}
export class InvalidFileExtensionError extends Error {}
export class EmptyFilePathError extends Error {}

const ZIP_EXT_CHECK_REGEX = /\.zip$/;
let progress;
const MONTHS = [
  'JANUARY',
  'FEBRUARY',
  'MARCH',
  'APRIL',
  'MAY',
  'JUNE',
  'JULY',
  'AUGUST',
  'SEPTEMBER',
  'OCTOBER',
  'NOVEMBER',
  'DECEMBER',
];
/**
 * Safe paths is interested in locations for latest a couple of weeks.
 * Date for latest 2 months should be sufficient to cover all cases,
 * especially the case when we are in the early days of the current month.
 * @returns {string[]} - array of files for latest 2 months from google takeout archive
 */
export function getFilenamesForLatest2Months(rootPath, now) {
  const previousMonth = dayjs(now).subtract(1, 'month');

  return [previousMonth, now].map(date => {
    const year = date.year();
    const monthStr = MONTHS[date.month()];
    return (
      `${rootPath}/Takeout/Location History/Semantic Location History/${year}/` +
      `${year}_${monthStr}.json`
    );
  });
}

// Imports any Takeout location data
// Currently works for Google Takeout Location data
export async function importTakeoutData(filePath) {
  if (!filePath) {
    throw new EmptyFilePathError();
  }
  let unifiedPath = filePath;

  if (Platform.OS === 'ios') {
    unifiedPath = filePath.replace('file://', '');
  }

  if (!ZIP_EXT_CHECK_REGEX.test(unifiedPath)) {
    throw new InvalidFileExtensionError();
  }

  // UnZip Progress Bar Log.
  // progress callback is required by unzip().
  progress = subscribe(
    ({
      progress,
      //  unifiedPath
    }) => {
      if (Math.trunc(progress * 100) % 10 === 0)
        console.log('[INFO] Unzipping', Math.trunc(progress * 100), '%');
    },
  );

  const extractDir = `${
    RNFS.CachesDirectoryPath
  }/Takeout-${new Date().toISOString()}`;

  console.log('[INFO] Takeout import start. Path:', unifiedPath);

  let path;
  let parsedFilesCount = 0;
  try {
    path = await unzip(unifiedPath, extractDir);

    console.log(`[INFO] Unzip Completed for ${path}`);

    const monthlyLocationFiles = getFilenamesForLatest2Months(path, dayjs());
    for (const filepath of monthlyLocationFiles) {
      console.log('[INFO] File to import:', filepath);

      const isExist = await RNFS.exists(`file://${filepath}`);
      if (isExist) {
        console.log('[INFO] File exists:', `file://${filepath}`);

        const contents = await RNFS.readFile(`file://${filepath}`).catch(() => {
          /**
           * IMPORTANT!!!
           * A temporary hack around URI generation bug in react-native-fs on android.
           * An exception is thrown as `file://` is not in the file URI:
           * "Error: ENOENT: No content provider:
           *  /data/user/0/edu.mit.privatekit/cache/Takeout-2020-04-12T15:48:54.295Z/Takeout/Location History/Semantic Location History/2020/2020_APRIL.json,
           *  open '/data/user/0/edu.mit.privatekit/cache/Takeout-2020-04-12T15:48:54.295Z/Takeout/Location History/Semantic Location History/2020/2020_APRIL.json'
           * "
           * @see https://github.com/itinance/react-native-fs/blob/master/android/src/main/java/com/rnfs/RNFSManager.java#L110
           */
          return RNFS.readFile(`file://file://${filepath}`);
        });

        let googleLocations = extractLocations(JSON.parse(contents));
        await NativeModules.SecureStorageManager.importGoogleLocations(
          googleLocations,
        );
        console.log('[INFO] Imported file:', filepath);
        parsedFilesCount++;
      }
    }
  } catch (err) {
    console.log('[Error] Failed to import Google Takeout', err);
  }
  // clean up unzipped folders
  if (path) {
    await RNFS.unlink(path);
  }
  progress.remove();
  if (parsedFilesCount === 0) {
    throw new NoRecentLocationsError();
  }
}
