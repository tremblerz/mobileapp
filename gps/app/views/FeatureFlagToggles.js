/* eslint-disable react-native/no-raw-text */
import React, { useEffect, useState } from 'react';
import { BackHandler, StyleSheet, View } from 'react-native';

import NavigationBarWrapper from '../components/NavigationBarWrapper';
import { Switch } from '../components/Switch';
import { Typography } from '../components/Typography';
import Colors from '../constants/colors';
import { getBuildtimeFlags } from '../constants/flagsEnv';
import { DEBUG_MODE } from '../constants/storage';
import { Theme } from '../constants/themes';
import { getCleanedFlagName, useFlags } from '../helpers/Flags';
import { GetStoreData } from '../helpers/General';
import { disableDebugMode, enableDebugMode } from '../helpers/Intersect';

export const FEATURE_FLAG_SCREEN_NAME = 'FeatureFlagsScreen';

export const FlagToggleRow = ({ name, val, onValueChange }) => {
  const buildtimeFlags = getBuildtimeFlags();

  const getBuildtimeFlagVal = flagName =>
    buildtimeFlags[flagName] ? 'On' : 'Off';

  return (
    <View style={styles.toggleRow}>
      <Switch
        testID={name}
        value={val}
        onValueChange={() => onValueChange(name)}
      />
      <View style={styles.toggleRowText}>
        <Typography use='body1' bold>
          {`${getCleanedFlagName(name)}`}
        </Typography>
        <Typography use='body2'>
          Default:
          <Typography style={styles.toggleDefaultVal} use='body3'>
            {` ${getBuildtimeFlagVal(name)}`}
          </Typography>
        </Typography>
      </View>
    </View>
  );
};

export const ExposureModeToggleRow = () => {
  const [isExposed, setIsExposed] = useState(false);

  useEffect(() => {
    const setInitalExposureVal = async () => {
      const isExposed = await GetStoreData(DEBUG_MODE, false);
      setIsExposed(isExposed);
    };
    setInitalExposureVal();
  }, []);

  const toggleExposureMode = val => {
    val ? enableDebugMode() : disableDebugMode();
    setIsExposed(val);
  };

  return (
    <View style={styles.toggleRow}>
      <Switch
        value={isExposed}
        onValueChange={toggleExposureMode}
        trackColor={{ false: Colors.GRAY_BACKGROUND, true: Colors.RED_TEXT }}
      />
      <Typography
        use='body1'
        bold
        style={
          isExposed
            ? { ...styles.toggleRowText, color: Colors.RED_TEXT }
            : styles.toggleRowText
        }>
        Exposure Mode
      </Typography>
    </View>
  );
};

export const FlagToggleList = () => {
  const [flags, setFlag] = useFlags();
  const alphabetizedFlags = Object.keys(flags).sort(); // Prevents toggle reordering

  const toggleFlag = async key => setFlag(key, !flags[key]);

  return (
    <View style={styles.toggleList}>
      <ExposureModeToggleRow />
      {alphabetizedFlags.map(key => {
        return (
          <FlagToggleRow
            key={key}
            name={key}
            val={flags[key]}
            onValueChange={toggleFlag}
          />
        );
      })}
    </View>
  );
};

export const FeatureFlagsScreen = ({ navigation }) => {
  useEffect(() => {
    const handleBackPress = () => {
      navigation.goBack();
      return true;
    };

    BackHandler.addEventListener('hardwareBackPress', handleBackPress);

    return () => {
      BackHandler.removeEventListener('hardwareBackPress', handleBackPress);
    };
  }, [navigation]);

  const backToMain = () => {
    navigation.goBack();
  };

  return (
    <NavigationBarWrapper title={'Feature Flags'} onBackPress={backToMain}>
      <Theme use='default'>
        <Typography use='body1' style={styles.headerText} bold>
          Notice
        </Typography>
        <Typography use='body2' style={styles.headerText}>
          This screen is only available when running the application in
          development or testing
        </Typography>
        <FlagToggleList />
      </Theme>
    </NavigationBarWrapper>
  );
};

const styles = StyleSheet.create({
  headerText: {
    textAlign: 'center',
    paddingTop: 7.5,
    marginHorizontal: 15,
  },
  toggleList: {
    paddingTop: 20,
    marginHorizontal: 10,
    width: '80%',
  },
  toggleRow: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 10,
    paddingHorizontal: 5,
    justifyContent: 'flex-start',
  },
  toggleRowText: {
    paddingLeft: 15,
  },
  toggleDefaultVal: {
    fontStyle: 'italic',
  },
});
