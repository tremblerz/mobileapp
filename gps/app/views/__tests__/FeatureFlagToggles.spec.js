import { act, fireEvent, render, wait } from '@testing-library/react-native';
import React from 'react';

import { FlagsProvider } from '../../helpers/Flags';
import { FeatureFlagsScreen } from '../FeatureFlagToggles';

jest.mock('../../helpers/General', () => {
  return {
    GetStoreData: jest.fn().mockResolvedValue(false),
  };
});

it('renders default values from the flags provider', async () => {
  const { asJSON } = render(
    <FlagsProvider>
      <FeatureFlagsScreen />
    </FlagsProvider>,
  );

  expect(asJSON()).toMatchSnapshot();
});

// TODO: issue in test lib: https://github.com/testing-library/native-testing-library/issues/46
// eslint-disable-next-line jest/no-disabled-tests
it.skip('renders toggle events', async () => {
  const { getByTestId } = render(
    <FlagsProvider>
      <FeatureFlagsScreen />
    </FlagsProvider>,
  );

  const toggle = getByTestId('feature1');

  await act(() => {
    fireEvent.valueChange(toggle, false);
  });

  await wait();

  expect(toggle.props.value).toBe(false);
});
