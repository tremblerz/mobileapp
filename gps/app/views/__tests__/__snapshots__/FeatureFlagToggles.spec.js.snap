// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default values from the flags provider 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#4051DB",
        "flexBasis": 0,
        "flexGrow": 0,
        "flexShrink": 1,
      }
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "#F7F8FF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#4051DB",
          "borderBottomColor": "#5061E6",
          "borderBottomWidth": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        isTVSelectable={true}
        style={
          Object {
            "alignItems": "center",
            "height": 55,
            "justifyContent": "center",
            "opacity": 1,
            "width": 60,
            "zIndex": 1,
          }
        }
      >
        <SvgXml
          style={
            Object {
              "color": "#FFF",
              "height": 18,
              "opacity": 0.4,
              "width": 18,
            }
          }
          xml="
  <svg width=\\"23\\" height=\\"24\\" viewBox=\\"0 0 23 24\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\">
    <path d=\\"M11.4854 3L2.22803 12.2573L11.4854 21.5146\\" stroke=\\"currentColor\\" stroke-opacity=\\"0.4\\" stroke-width=\\"3\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" />
  </svg>
"
        />
      </View>
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#FFF",
            "fontFamily": "IBMPlexSans-Medium",
            "fontSize": 26,
            "letterSpacing": -0.3,
            "lineHeight": 34,
            "paddingHorizontal": 20,
            "position": "absolute",
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Feature Flags
      </Text>
    </View>
    <Text
      bold={true}
      style={
        Object {
          "color": "#4051DB",
          "fontFamily": "IBMPlexSans-Bold",
          "fontSize": 18,
          "fontWeight": "bold",
          "lineHeight": 24,
          "marginHorizontal": 15,
          "paddingTop": 7.5,
          "textAlign": "center",
          "writingDirection": "ltr",
        }
      }
      use="body1"
    >
      Notice
    </Text>
    <Text
      style={
        Object {
          "color": "#4051DB",
          "fontFamily": "IBMPlexSans",
          "fontSize": 16,
          "fontWeight": "normal",
          "lineHeight": 22,
          "marginHorizontal": 15,
          "paddingTop": 7.5,
          "textAlign": "center",
          "writingDirection": "ltr",
        }
      }
      use="body2"
    >
      This screen is only available when running the application in development or testing
    </Text>
    <View
      style={
        Object {
          "marginHorizontal": 10,
          "paddingTop": 20,
          "width": "80%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "paddingHorizontal": 5,
            "paddingVertical": 10,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="button"
          onTintColor="#FF5656"
          style={
            Object {
              "height": 31,
              "width": 51,
            }
          }
          testID="switch"
          thumbTintColor="#FFF"
          tintColor="#DADADA"
          value={false}
        />
        <Text
          bold={true}
          style={
            Object {
              "color": "#4051DB",
              "fontFamily": "IBMPlexSans-Bold",
              "fontSize": 18,
              "fontWeight": "bold",
              "lineHeight": 24,
              "paddingLeft": 15,
              "writingDirection": "ltr",
            }
          }
          use="body1"
        >
          Exposure Mode
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "paddingHorizontal": 5,
            "paddingVertical": 10,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="button"
          onTintColor="#41dca4"
          style={
            Object {
              "height": 31,
              "width": 51,
            }
          }
          testID="FOO_BAR"
          thumbTintColor="#FFF"
          tintColor="#DADADA"
          value={true}
        />
        <View
          style={
            Object {
              "paddingLeft": 15,
            }
          }
        >
          <Text
            bold={true}
            style={
              Object {
                "color": "#4051DB",
                "fontFamily": "IBMPlexSans-Bold",
                "fontSize": 18,
                "fontWeight": "bold",
                "lineHeight": 24,
                "writingDirection": "ltr",
              }
            }
            use="body1"
          >
            FOO BAR
          </Text>
          <Text
            style={
              Object {
                "color": "#4051DB",
                "fontFamily": "IBMPlexSans",
                "fontSize": 16,
                "fontWeight": "normal",
                "lineHeight": 22,
                "writingDirection": "ltr",
              }
            }
            use="body2"
          >
            Default:
            <Text
              style={
                Object {
                  "color": "#4051DB",
                  "fontFamily": "IBMPlexSans",
                  "fontSize": 15,
                  "fontStyle": "italic",
                  "fontWeight": "normal",
                  "lineHeight": 24,
                  "writingDirection": "ltr",
                }
              }
              use="body3"
            >
               On
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
